### PR TITLE
refactor flaw injection

### LINKED ISSUE #1409: BUILD_FLAW_LEVELS error on broken_links and bad_bcd_queries always triggered
For example:
```
▶ BUILD_FOLDERSEARCH=glossary/png BUILD_FLAW_LEVELS="*:ignore, broken_links: error" yarn build
yarn run v1.22.4
$ cd build && cross-env NODE_ENV=production node cli.js
Building roots: [ '/Users/peterbe/dev/MOZILLA/MDN/content/files' ]
error while building documents: Error: broken_links flaws: undefined
    at injectFlaws (/Users/peterbe/dev/MOZILLA/MDN/yari/build/flaws.js:133:13)
    at buildDocument (/Users/peterbe/dev/MOZILLA/MDN/yari/build/index.js:245:3)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
    at async buildDocuments (/Users/peterbe/dev/MOZILLA/MDN/yari/build/cli.js:65:59)
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```
(note the `BUILD_FLAW_LEVELS="*:ignore, broken_links: error"`)

I.e. if the level is set to `error` it will throw even if there were 0 `broken_links` flaws. 

Equally, with the `bad_bcd_queries`, but due to a bug, it actually has `broken_links` where it was meant to be `bad_bcd_queries`. See this line: https://github.com/mdn/yari/blob/f653aee636cc89243ffe695d4d01fbd729db92a9/build/flaws.js#L164
Even without that typo, the logic is wrong. It throws an error independent of the number of actual `bad_bcd_queries`.

### PR DESCRIPTION
**Pull request by peterbe**
Fixes #1409

First of all, the diff looks huge but it's just unfortunate from the refactoring. 
It used to be this, all in one big function:

```javascript
function injectFlaws() {
  if (options.flawLevels.get("broken_links") !== FLAW_LEVELS.IGNORE) {
    ...
  }
  if (options.flawLevels.get("bad_bcd_queries") !== FLAW_LEVELS.IGNORE) {
    ...
  }
}
```

And, apart from the actual important fixes, my refactoring was to break it into smaller functions:

```javascript
function injectFlaws() {
  injectBrokenLinksFlaws(options.flawLevels.get("broken_links"), ...);
  injectBadBCDQueriesFlaws(options.flawLevels.get("bad_bcd_queries"), ...);
}
```
and within those broken-out functions, I made it return early if `level === FLAW_LEVELS.IGNORE`.

Hence the diff looking so large.

### CODE REVIEW

**Comment by peterbe** [comment]
To test it, try:
```
▶ BUILD_FOLDERSEARCH=glossary/png BUILD_FLAW_LEVELS="*:ignore, bad_bcd_queries: error, broken_links: error" yarn build
...
✨  Done in 2.74s.
```
See, even with the error levels turned up to 11, it doesn't crash if the pages it builds are perfectly flawless.

**Comment by peterbe** [comment]
The other thing I wanted to fix was, that if a flaw error is thrown, it's not at all clear which document it failed on. So you'd get:
```
error while building documents: Error: bad_bcd_queries flaws: No BCD data for query: webextensions.api.bookmarks.export
    at injectBadBCDQueriesFlaws (/Users/peterbe/dev/MOZILLA/MDN/yari/build/flaws.js:195:11)
    at injectFlaws (/Users/peterbe/dev/MOZILLA/MDN/yari/build/flaws.js:25:3)
    at buildDocument (/Users/peterbe/dev/MOZILLA/MDN/yari/build/index.js:246:5)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
    at async buildDocuments (/Users/peterbe/dev/MOZILLA/MDN/yari/build/cli.js:65:59)
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```
which, if you're building all document's doesn't get you much hope to single out which document is the problem. So I added the extra `try/catch` and now it spits this out instead:
```
Injecting flaws into /Users/peterbe/dev/MOZILLA/MDN/content/files/en-us/mozilla/add-ons/webextensions/api/bookmarks/export/index.html (/en-US/docs/Mozilla/Add-ons/WebExtensions/API/bookmarks/export) failed.
error while building documents: Error: bad_bcd_queries flaws: No BCD data for query: webextensions.api.bookmarks.export
    at injectBadBCDQueriesFlaws (/Users/peterbe/dev/MOZILLA/MDN/yari/build/flaws.js:195:11)
    at injectFlaws (/Users/peterbe/dev/MOZILLA/MDN/yari/build/flaws.js:25:3)
    at buildDocument (/Users/peterbe/dev/MOZILLA/MDN/yari/build/index.js:246:5)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
    at async buildDocuments (/Users/peterbe/dev/MOZILLA/MDN/yari/build/cli.js:65:59)
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

File: build/flaws.js [hunk 1]
```diff
@@ -15,159 +15,188 @@ const {
...
-          }
-          doc.flaws.bad_bcd_queries.push({
-            id: `bad_bcd_queries${doc.flaws.bad_bcd_queries.length}`,
-            explanation: `No BCD data for query: ${query}`,
-            suggestion: null,
-          });
-        }
       }
-    });
-    if (options.flawLevels.get("broken_links") === FLAW_LEVELS.ERROR) {
```
Comment (peterbe) [inline build/flaws.js:164]: This was the old bug. It looked to see if the flaw level for `broken_links` was set to error in the context of checking `bad_bcd_queries`.

File: build/flaws.js [hunk 2]
```diff
@@ -15,159 +15,188 @@ const {
...
-        `bad_bcd_queries flaws: ${doc.flaws.bad_bcd_queries.map(
-          (f) => f.explanation
-        )}`
-      );
     }
+  });
+  if (
+    level === FLAW_LEVELS.ERROR &&
+    doc.flaws.bad_bcd_queries &&
+    doc.flaws.bad_bcd_queries.length
```
Comment (peterbe) [inline build/flaws.js:193]: These two lines are important.

File: build/flaws.js [hunk 3]
```diff
@@ -15,159 +15,188 @@ const {
...
       }
-    });
-    if (options.flawLevels.get("broken_links") === FLAW_LEVELS.ERROR) {
-      throw new Error(`broken_links flaws: ${doc.flaws.broken_links}`);
     }
+  });
+  if (
+    level === FLAW_LEVELS.ERROR &&
+    doc.flaws.broken_links &&
+    doc.flaws.broken_links.length
```
Comment (peterbe) [inline build/flaws.js:147]: These two lines are important.

**Comment by peterbe** [comment]
I'm aware that the recent work on resolving redirects is changing in the [fundamental redirects PR](https://github.com/mdn/yari/pull/1359) so let's let that finish first and I'll fix the merge conflicts before merging *this* PR. However, I think it can be reviewed whenever.

**Review by escattone — APPROVED** [review]
Thanks @peterbe, I like these improvements.

**Review by escattone — APPROVED** [review]

**Pull request merged**